### PR TITLE
feat: `Cache::set`

### PR DESCRIPTION
**Pull request by crepererum**
This will be helpful to fill caches if we got the information from
somewhere else.

For #3985.

### CODE REVIEW

File: querier/src/cache_system/driver.rs [hunk 1]
```diff
@@ -61,51 +64,80 @@ where
...
+                        // If we panic here then `tx` will be dropped and the receivers will be notified.
+                        let v = loader.load(k_for_loader).await;
+
+                        // remove "running" state and store result
+                        //
+                        // Note: we need to manually drop the result of `.remove(...).expect(...)` here to convince rustc
+                        //       that we don't need the shared future within the resulting tuple. The warning we would get
+                        //       is:
+                        //
+                        //       warning: unused `futures::future::Shared` in tuple element 0 that must be used
```
Comment (domodwyer) [inline querier/src/cache_system/driver.rs:98]: I bet this was fun...

**Review by domodwyer — APPROVED** [review]
Sync stuff looks good to me - I need to read up on what is going on here but I trust ✌️

File: querier/src/cache_system/driver.rs [hunk 2]
```diff
@@ -61,51 +64,80 @@ where
...
+                        v = loader_fut => v,
+                        maybe_v = rx_set => {
+                            match maybe_v {
+                                Ok(v) => {
+                                    // data get side-loaded via `Cache::set`. In this case, we do NOT modify the state
+                                    // because there would be a lock-gap. The `set` function will do that for us instead.
+                                    v
+                                }
+                                Err(_) => {
+                                    // sender side is gone, very likely the cache is shutting down
```
Comment (e-dard) [inline querier/src/cache_system/driver.rs:121]: Perhaps worth logging this error at a `debug` level?

File: querier/src/cache_system/driver.rs [hunk 3]
```diff
@@ -116,6 +148,22 @@ where
...
+    ///
+    /// This will also complete a currently running request for this key.
+    pub fn set(&self, k: K, v: V) {
+        let mut state = self.state.lock();
+
+        if let Some(running_query) = state.running_queries.remove(&k) {
+            // it's OK when the receiver side is gone (likely panicked)
+            running_query.set.send(v.clone()).ok();
+
+            // in this case, the task did NOT modify the backend, so we have to do that
```
Comment (e-dard) [inline querier/src/cache_system/driver.rs:162]: It's not clear to me which "case" this comment is referring to.

File: querier/src/cache_system/driver.rs [hunk 4]
```diff
@@ -144,6 +192,21 @@ where
 /// - `Shared`: Allow the receiver to be cloned and be awaited from multiple places.
 type SharedReceiver<V> = Shared<BoxFuture<'static, Result<Arc<Mutex<V>>, Arc<RecvError>>>>;
 
+/// Running query
```
Comment (e-dard) [inline querier/src/cache_system/driver.rs:195]: How about something like:

```suggestion
/// State for coordinating the execution of a single running query.
```

or something?

**Review by e-dard — APPROVED** [review]
Good stuff 👍

File: querier/src/cache_system/driver.rs (hunk 3 above)
Comment (crepererum) [reply querier/src/cache_system/driver.rs]: Extended the comment.

File: querier/src/cache_system/driver.rs (hunk 2 above)
Comment (crepererum) [reply querier/src/cache_system/driver.rs:122]: done

**Pull request merged**